src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompeltionListUpdater.cs

```diff
@@ -65,26 +82,69 @@ private sealed class CompletionListUpdater
 
-                // Let us make the completion Helper used for non-Roslyn items case-sensitive.
-                // We can change this if get requests from partner teams.
-                DefaultCompletionHelper = new CompletionHelper(isCaseSensitive: true);
-            }
+                FilterText = Session.ApplicableToSpan.GetText(Data.Snapshot);
+                InitialRoslynTriggerKind = Helpers.GetRoslynTriggerKind(Data.InitialTrigger);
 
-            public FilteredCompletionModel? UpdateCompletionList()
-            {
                 if (!Session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
                 {
                     // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
                     // For now, the default hasSuggestedItemOptions is false.
                     hasSuggestedItemOptions = false;
                 }
 
-                hasSuggestedItemOptions |= Data.DisplaySuggestionItem;
+                HasSuggestedItemOptions = hasSuggestedItemOptions || Data.DisplaySuggestionItem;
+
+                // We prefer using the original snapshot, which should always be available from items provided by Roslyn's CompletionSource.
+                // Only use data.Snapshot in the theoretically possible but rare case when all items we are handling are from some non-Roslyn CompletionSource.
+                var snapshotForDocument = TryGetInitialTriggerLocation(Data, out var intialTriggerLocation)
+                    ? intialTriggerLocation.Snapshot
+                    : Data.Snapshot;
 
-                var filterText = Session.ApplicableToSpan.GetText(Data.Snapshot);
-                var reason = Data.Trigger.Reason;
-                var initialRoslynTriggerKind = Helpers.GetRoslynTriggerKind(Data.InitialTrigger);
+                Document = snapshotForDocument?.TextBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+                CompletionService = Document?.GetLanguageService<CompletionService>();
+                CompletionRules = CompletionService?.GetRules(CompletionOptions.From(Document!.Project)) ?? CompletionRules.Default;
 
+                // Let us make the completion Helper used for non-Roslyn items case-sensitive.
+                // We can change this if get requests from partner teams.
+                CompletionHelper = Document != null ? CompletionHelper.GetHelper(Document) : new CompletionHelper(isCaseSensitive: true);
+
+                // Nothing to highlight if user hasn't typed anything yet.
+                HighlightMatchingPortions = FilterText.Length > 0
+                    && GlobalOptions.GetOption(CompletionViewOptions.HighlightMatchingPortionsOfCompletionListItems, Document?.Project.Language);
+
+                ItemsToBeIncluded = s_listOfMatchResultPool.Allocate();
+            }
+
+            public FilteredCompletionModel? UpdateCompletionList()
```

Comment:
just call method `Filter()`?

Reply:
Not just filtering though, this helper class handles filtering, sorting and selection, based on user inputs (i.e. typed chars, selected filters, etc.)